ruby/ext/google/protobuf_c/defs.c

```diff
@@ -1294,6 +1302,308 @@ static void EnumDescriptor_register(VALUE module) {
+} ServiceDescriptor;
+
+static VALUE cServiceDescriptor = Qnil;
+
+static void ServiceDescriptor_mark(void* _self) {
+  ServiceDescriptor* self = _self;
+  rb_gc_mark(self->module);
+  rb_gc_mark(self->descriptor_pool);
+}
+
+static const rb_data_type_t ServiceDescriptor_type = {
+    "Google::Protobuf::ServicDescriptor",
+    {ServiceDescriptor_mark, RUBY_DEFAULT_FREE, NULL},
+    .flags = RUBY_TYPED_FREE_IMMEDIATELY | RUBY_TYPED_WB_PROTECTED,
+};
+
+static ServiceDescriptor* ruby_to_ServiceDescriptor(VALUE val) {
+  ServiceDescriptor* ret;
+  TypedData_Get_Struct(val, ServiceDescriptor, &ServiceDescriptor_type, ret);
+  return ret;
+}
+
+static VALUE ServiceDescriptor_alloc(VALUE klass) {
+  ServiceDescriptor* self = ALLOC(ServiceDescriptor);
+  VALUE ret = TypedData_Wrap_Struct(klass, &ServiceDescriptor_type, self);
+  self->servicedef = NULL;
+  self->module = Qnil;
+  self->descriptor_pool = Qnil;
+  return ret;
+}
+
+/*
+ * call-seq:
+ *    ServiceDescriptor.new(c_only_cookie, ptr) => ServiceDescriptor
+ *
+ * Creates a descriptor wrapper object.  May only be called from C.
+ */
+static VALUE ServiceDescriptor_initialize(VALUE _self, VALUE cookie,
+                                       VALUE descriptor_pool, VALUE ptr) {
+  ServiceDescriptor* self = ruby_to_ServiceDescriptor(_self);
+
+  if (cookie != c_only_cookie) {
+    rb_raise(rb_eRuntimeError,
+             "Descriptor objects may not be created from Ruby.");
+  }
+
+  RB_OBJ_WRITE(_self, &self->descriptor_pool, descriptor_pool);
+  self->servicedef = (const upb_ServiceDef*)NUM2ULL(ptr);
+
+  return Qnil;
+}
+
+/*
+ * call-seq:
+ *     ServiceDescriptor.name => name
+ *
+ * Returns the name of this service.
+ */
+static VALUE ServiceDescriptor_name(VALUE _self) {
+  ServiceDescriptor* self = ruby_to_ServiceDescriptor(_self);
+  return rb_str_maybe_null(upb_ServiceDef_FullName(self->servicedef));
+}
+
+/*
+ * call-seq:
+ *    ServiceDescriptor.file_descriptor
+ *
+ * Returns the FileDescriptor object this service belongs to.
+ */
+static VALUE ServiceDescriptor_file_descriptor(VALUE _self) {
+  ServiceDescriptor* self = ruby_to_ServiceDescriptor(_self);
+  return get_filedef_obj(self->descriptor_pool,
+                         upb_ServiceDef_File(self->servicedef));
+}
+
+
+/*
+ * call-seq:
+ *     ServiceDescriptor.each_method(&block)
+ *
+ * Iterates over methods in this service, yielding to the block on each one.
+ */
+static VALUE ServiceDescriptor_each_method(VALUE _self) {
+  ServiceDescriptor* self = ruby_to_ServiceDescriptor(_self);
+
+  int n = upb_ServiceDef_MethodCount(self->servicedef);
+  for (int i = 0; i < n; i++) {
+    const upb_MethodDef *method = upb_ServiceDef_Method(self->servicedef, i);
+    VALUE obj = get_methoddef_obj(self->descriptor_pool, method);
+    rb_yield(obj);
+  }
+  return Qnil;
+}
+
+/*
+ * call-seq:
+ *    ServiceDescriptor.methods => Enumerator
+ *
+ * Returns an enumerator over all the methods in this service
```

Comment:
> If we want to add methods returning enumerators later, let's do it consistently across all the descriptor types.

So some of our other descriptor types (including plain message `Descriptor`) _are_ enumerables though (they include the Enumerable mixin), which is more or less the same thing. So you can do

```
descriptor.each { |field| puts "have field #{field}" }
```

but also things like

```
descriptor.to_a
# or
descriptor.select { |field| field.type == 'int32' }
```

I didn't implement `ServiceDescriptor#each`, however, because I don't think a service is-a collection of RPC methods in the same way that a message type is-a collection of fields. So instead, I implemented `#each_method`, and thus implemented `#methods` to return an enumerable in the same way that `self` is an enumerator for a message descriptor.

If you like I could just implement `#each_method` as `ServiecDescriptor#each` instead, if you think service descriptor is-a list of RPC methods. (as a bonus that would solve the `#methods` method shadowing `Object#methods` too, since we wouldn't need that since `self` _is_ the enumerable then.

Reply:
I see your point, that "service is-a collection of methods" is a bit of a stretch. But methods *are* the only thing you can list inside a `service {}` block in a `.proto` file, other than options, which have their own accessor. The fact that this conveniently resolves the `#methods` shadowing is another argument in its favor.

Reply:
alright 👍 the current version of my PR just implements `#each`/`Enumerable` directly on `ServiceDescriptor`. Thanks for talking through that with me.